#12273 Fix twisted.conch tests with OpenSSH >= 9.8

cjwatson opened the pull request:
## Scope and purpose

Fixes #12273

OpenSSH 9.8 disables support for DSA keys by default at compile-time (https://www.openssh.com/releasenotes.html#9.8p1).  Since DSA keys are now largely a thing of the past, it seems unwise for Twisted Conch's tests to rely on them any more.  These tests aren't testing anything specifically about DSA, and an RSA key will do just as well.

Unit tests for Twisted Conch's DSA features still exist; those tests don't rely on support in OpenSSH.


## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

cjwatson committed:
Fix twisted.conch tests with OpenSSH >= 9.8

OpenSSH 9.8 disables support for DSA keys by default at compile-time
(https://www.openssh.com/releasenotes.html#9.8p1).  Since DSA keys are
now largely a thing of the past, it seems unwise for Twisted Conch's
tests to rely on them any more.  These tests aren't testing anything
specifically about DSA, and an RSA key will do just as well.

Unit tests for Twisted Conch's DSA features still exist; those tests
don't rely on support in OpenSSH.

cjwatson commented on src/twisted/conch/test/test_cftp.py:
```diff
@@ -1475,36 +1475,25 @@ def _getAttrs(self, s):
+            fn,
+            "testuser@127.0.0.1",
+        )
+        return args
```
This was a mistake (and causes the slight reduction in coverage), but when I remove this line the test tries to use my normal keys and prompts me for a passphrase, which is clearly wrong.  I'll figure this out soon.

adiroiban commented on src/twisted/conch/test/test_conch.py:
```diff
@@ -547,51 +536,34 @@ def execute(self, remoteCommand, process, sshArgs=""):
+        cmds = (cmdline % port).split()
+        encodedCmds = []
+        for cmd in cmds:
+            if isinstance(cmd, str):
+                cmd = cmd.encode("utf-8")
```
```suggestion
            cmd = cmd.encode("utf-8")
```

It looks like there is a missing branch coverage here

![image](https://github.com/user-attachments/assets/d91b2810-7620-42c2-9482-2e6326e79dd8)

cjwatson replied:
I did something similar, thanks.

adiroiban reviewed (approved):
It looks good.

Only a minor comment about missing branch coverage.

You can fix and then merge.


Basically this is switching the OpenSSH integration tests from using DSA to RSA keys.

twm commented:
If it's not stable it's not a useful perf test. I've filed #12276 for this and plan to take a look in the next few days. In the meantime I've ack'd the regression here.

cjwatson committed:
Run ssh/sftp with -oIdentitiesOnly=yes

When running locally in an environment where an `ssh-agent` process is
running, this stops these tests trying to use identities from it.

cjwatson committed:
Simplify OpenSSHClientMixin.execute slightly

This improves coverage.
